src/messages/validate.rs

```diff
@@ -0,0 +1,145 @@
+}
+
+impl Validate for Message {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        match self.payload {
+            Payload::SharePackage(_) => {
+                if self.header.sender != ParticipantId::Dealer {
+                    return Err(MsgErr::SenderMustBeDealer);
+                }
+                if !matches!(self.header.receiver, ParticipantId::Signer(_)) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
+            }
+            Payload::SigningCommitments(_) => {
+                if !matches!(self.header.sender, ParticipantId::Signer(_)) {
+                    return Err(MsgErr::SenderMustBeSigner);
+                }
+                if self.header.receiver != ParticipantId::Aggregator {
+                    return Err(MsgErr::ReceiverMustBeAggergator);
+                }
+            }
+            Payload::SigningPackage(_) => {
+                if self.header.sender != ParticipantId::Aggregator {
+                    return Err(MsgErr::SenderMustBeAggregator);
+                }
+                if !matches!(self.header.receiver, ParticipantId::Signer(_)) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
+            }
+            Payload::SignatureShare(_) => {
+                if !matches!(self.header.sender, ParticipantId::Signer(_)) {
+                    return Err(MsgErr::SenderMustBeSigner);
+                }
+                if self.header.receiver != ParticipantId::Aggregator {
+                    return Err(MsgErr::ReceiverMustBeAggergator);
+                }
+            }
+            Payload::AggregateSignature(_) => {
+                if self.header.sender != ParticipantId::Aggregator {
+                    return Err(MsgErr::SenderMustBeAggregator);
+                }
+                if !matches!(self.header.receiver, ParticipantId::Signer(_)) {
+                    return Err(MsgErr::ReceiverMustBeSigner);
+                }
+            }
+        }
+
+        Ok(self)
+    }
+}
+
+impl Validate for Header {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        // Validate the message version.
+        // By now we only have 1 valid version so we compare against that.
+        if self.version != BASIC_FROST_SERIALIZATION {
+            return Err(MsgErr::WrongVersion);
+        }
+
+        // Make sure the sender and the receiver are not the same.
+        if self.sender == self.receiver {
+            return Err(MsgErr::SameSenderAndReceiver);
+        }
+        Ok(self)
+    }
+}
+
+impl Validate for Payload {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        match self {
+            Payload::SharePackage(share_package) => {
+                if share_package.share_commitment.len() < MIN_SIGNERS {
+                    return Err(MsgErr::NotEnoughCommitments(MIN_SIGNERS));
+                }
+
+                if share_package.share_commitment.len() > MAX_SIGNER_PARTICIPANT_ID.into() {
```

Comment:
`MAX_SIGNER_PARTICIPANT_ID` is 253, but the maximum number of signers is actually 254.
(We reserve 2/256 IDs for the dealer and aggregator, leaving 254 valid IDs.)

Let's define a separate constant:
```
pub const MAX_SIGNERS: u8 = MAX_SIGNER_PARTICIPANT_ID + 1;
```

Then make this fix:
```suggestion
                if share_package.share_commitment.len() > MAX_SIGNERS.into() {
```